### ISSUE TITLE
add support for include in lists

### ISSUE THREAD

**Issue by diegoiast** [diegoiast/qutepart-cpp#10]
List can be verbatim, or include other lists. This broke ZSH support, and is currently bypassed. See 
75202e58687bc94805ab3722887a8ff006f77e75 for more details.

**Comment by diegoiast** [comment]
I instructed claude to get the upstream code and ours - and make list of missing features. This is in hope that this qualifies as a proper black box implementation (even though as always the LLM is way too descriptive and very verbose) . The document shows some minor code snippets from kate, but I don't think this is a "huge problem". 

Worth noting is that the document describes spelling, and this is being worked on (see https://github.com/diegoiast/qutepart-cpp/issues/60). Code seems to work on Linux, but on Windows it will need to download the language dictionaries from somewhere (at runtime? bundle them inside the calling application?) and instruct the low level unspell to use them. macOS should work out of the box (since it does use OS native APIs) but is untested.

This is what it spited:

# What qutepart-cpp is missing from Kate's highlighting format (2026)

Status: written 2026-08-25 against qutepart-cpp `7f5bec9`, compared to
KSyntaxHighlighting `4c17516` (2026-08-25) — a full clone, so every claim below
about upstream behaviour is checked against its source, not just against
`data/schema/language.xsd`.

## Scope and method

qutepart-cpp reads the same `<language>` XML that KSyntaxHighlighting (the
library behind KatePart) reads, and the same `.theme` JSON. This document lists
what the qutepart-cpp loader does **not** do, feature by feature.

Every gap below was checked three ways:

1. against the upstream XSD (which elements/attributes *may* exist),
2. against the upstream C++ (what upstream actually *reads* and *does* — the
   XSD lists several attributes upstream itself ignores), and
3. against the 385 syntax files bundled in `syntax/`, to count how many real
   language definitions depend on the feature.

The "affected" numbers are counts over the **bundled** files. They are a floor,
not a ceiling: the bundled data is itself several years behind upstream (see
§6), and newer upstream definitions lean harder on the newer attributes.

**Good news first.** All 18 rule types exist and all 31 `dsXxx` default styles
are recognised. There is no missing rule *kind*. Everything below is either a
missing attribute, a missing `<general>` sub-element, a missing subsystem
(indentation folding, spell-check regions), or an attribute that is parsed and
then dropped on the floor.

---

## Priority summary

| # | Gap | Kind | Bundled files affected |
|---|-----|------|------------------------|
| 1.1 | `fallthroughContext` ignored unless deprecated `fallthrough="true"` is also present | correctness | 28 files, 399 contexts |
| 1.2 | `StringDetect insensitive` parsed then ignored | correctness | 46 files, 302 rules |
| 1.3 | `StringDetect dynamic` regex-escapes its substitution | correctness | 21 files, 56 rules |
| 1.4 | Dynamic `DetectChar` reads the whole match instead of capture group 1 | correctness | 2 files, 8 rules |
| 2.1 | Indentation-based folding (`<folding indentationsensitive>`) | missing subsystem | 34 files |
| 2.2 | `noIndentationBasedFolding` on `<context>` | missing attribute | 10 files (blocked on 2.1) |
| 3.1 | `<include>` inside a keyword `<list>` | missing feature | 9 files |
| 3.2 | `IncludeRules includeAttrib` | missing attribute | 23 files |
| 3.3 | `keyword insensitive` (per-rule) | missing attribute | 8 files, 12 rules |
| 3.4 | `<language casesensitive>` | missing attribute | 51 files |
| 4.1 | `itemData backgroundColor` / `selBackgroundColor` / `selColor` | missing attribute | 61 files |
| 4.2 | Theme `custom-styles` parsed but never applied | dead code | all themes |
| 4.3 | `dsOperator` has no default format | incomplete | many |
| 5.1 | `<comment position>` | missing attribute | 30 files |
| 5.2 | `<emptyLines>` / `<emptyLine regexpr>` | missing element | 9 files |
| 5.3 | `<spellchecking><encodings>` and the `spellChecking` itemData flag | missing subsystem | 2 files + all |
| 5.4 | `<keywords wordWrapDeliminator>` | missing attribute | 9 files |
| 5.5 | `LineContinue char` | missing attribute | 7 files, 18 rules |
| 5.6 | Per-rule `weakDeliminator` / `additionalDeliminator` | missing attribute | 2 files |
| 5.7 | `<language>` metadata: `alternativeNames`, `section`, `author`, `license`, `style`, `generated` | missing attribute | 5+ files |
| 5.8 | `lineEmptyContext` switches once instead of looping; `stopEmptyLineContextSwitchLoop` unread | correctness | chained defs |
| 5.9 | Dynamic captures capped at `%4`; no two-digit support | incomplete | 5 rules |
| 5.10 | MIME types read from `mimetypes` (upstream: `mimetype`), then overwritten with globs | bug | 275 files |
| 6 | Syntax data 3–10 major revisions behind; 37 languages absent | data freshness | all |
| 7 | 7 of 15 indenter modes referenced by bundled files have no implementation | missing subsystem | 12 files |

**Checked and *not* gaps.** Some XSD attributes are unread by upstream too, so
ignoring them is correct: `<comment region>` (57 bundled files),
`<emptyLine casesensitive>`, and the `fallthrough` boolean itself. These are
called out inline where they arise, so nobody "fixes" them later.

---

# 1. Correctness gaps in features that *are* implemented

These are the ones to fix first. The feature is wired up, but it misbehaves, so
the affected languages highlight wrongly today with no diagnostic.

## 1.1 `fallthroughContext` is ignored without the deprecated `fallthrough` attribute

**What it does.** A context with `fallthroughContext="X"` means: if no rule in
this context matches at the current position, switch to `X` *instead of*
consuming one character as unmatched text. It is how Kate expresses "this
context only exists to recognise an optional prefix; if the prefix is not
there, move on". Without it, the parser eats a character, marks it with the
context's own attribute, and tries again — which usually mis-colours the rest
of the line.

Historically the XML needed both `fallthrough="true"` *and*
`fallthroughContext="X"`. Upstream has since dropped the boolean **entirely** —
`HighlightingContextData::load()`
(`upstream:src/lib/highlightingdata.cpp:336`) reads only `fallthroughContext`,
and `Context::fallthrough()` (`upstream:src/lib/context_p.h:49-52`) is simply:

```cpp
return !m_fallthroughContext.isStay();
```

So upstream's rule is: *fallthrough is on iff `fallthroughContext` is present
and does not resolve to `#stay`*. The `fallthrough` attribute is not read at
all, and no bundled file sets it to `"false"` (checked: 0 occurrences), so
there is nothing to preserve for backwards compatibility.

**What qutepart does.** `src/hl/loader.cpp:408-419`:

```cpp
bool fallthrough = parseBoolAttribute(getAttribute(attrs, "fallthrough", "false"), error);
...
if (fallthrough) {
    QString fallthroughContextText = getAttribute(attrs, "fallthroughContext", "#stay", true);
    fallthroughContext = makeContextSwitcher(fallthroughContextText);
}
```

If `fallthrough` is absent the `fallthroughContext` attribute is never read at
all. 399 contexts across 28 bundled files are silently degraded — worst
offenders `raku.xml` (120 contexts), `todo.xml` (58),
`logfile-advanced.xml` (34), `racket.xml` (31), `elvish.xml` (17),
`cabal.xml` (16), `fish.xml` (13).

**How to implement.** Match upstream: ignore the boolean, key off the context.

```cpp
ContextSwitcher fallthroughContext;
if (attrs.hasAttribute("fallthroughContext")) {
    fallthroughContext = makeContextSwitcher(getAttribute(attrs, "fallthroughContext"));
}
```

`makeContextSwitcher()` already returns a null `ContextSwitcher` for `#stay`,
which reproduces upstream's `!isStay()` test exactly — a `#stay` fallthrough
stays disabled, so the no-match path cannot spin.

The runtime side already works: `Context::parseBlock()`
(`src/hl/context.cpp:336`) switches on no-match when `fallthroughContext` is
non-null.

## 1.2 `StringDetect insensitive` is parsed and then ignored

**What it does.** `<StringDetect String="foo" insensitive="true"/>` matches
`FOO`, `Foo`, `fOo`. Used heavily by case-insensitive languages — SQL, Pascal,
Fortran, INI-likes, batch files.

**What qutepart does.** `loadStringRule()` (`src/hl/loader.cpp:161-176`) reads
the flag and hands it to `AbstractStringRule`, which stores it. But
`StringDetectRule::tryMatchImpl()` (`src/hl/rules.cpp:133-148`) never looks at
it:

```cpp
if (textToMatch.text.startsWith(pattern)) {   // always Qt::CaseSensitive
    return makeMatchResult(result, pattern.length());
}
```

Upstream honours it: every string matcher funnels through `matchString()`
(`upstream:src/lib/rule.cpp:98-104`), which takes a `Qt::CaseSensitivity`, and
`StringDetect::doMatch()` (`upstream:src/lib/rule.cpp:707-710`) passes its
`m_caseSensitivity`. qutepart's `WordDetectRule::tryMatchImpl()`
(`src/hl/rules.cpp:253`) does the same thing correctly — only `StringDetect`
was missed. 302 rules across 46 bundled files are affected.

**How to implement.** One-line fix in the matcher:

```cpp
auto cs = insensitive ? Qt::CaseInsensitive : Qt::CaseSensitive;
if (textToMatch.text.startsWith(pattern, cs)) {
    return makeMatchResult(result, pattern.length());
}
```

Then fix `StringDetectRule::computeFirstChars()` to seed **both** cases of the
first character when `insensitive` is set — otherwise the first-char bucket
index in `Context::buildRuleIndex()` will skip the rule for the column where it
would have matched, and the fix will appear not to work. Same applies to
`Detect2CharsRule` if insensitivity is ever added there.

## 1.3 `StringDetect dynamic` regex-escapes the substituted text

**What it does.** `dynamic="true"` substitutes capture groups from the match
that entered the current context into the rule's pattern. For `RegExpr` the
captures must be regex-escaped (a captured `.` should match a literal dot).
For `StringDetect` — a *literal* matcher — escaping is wrong: it injects
backslashes into the string being compared.

**What qutepart does.** `src/hl/rules.cpp:124-131`:

```cpp
QString makeDynamicSubsctitutions(QString pattern, const QStringList &data) {
    return pattern.replace("%0", QRegularExpression::escape(data.value(0)))
        ...
}
```

`StringDetectRule::tryMatchImpl()` calls this and then does a literal
`startsWith()`. A capture of `a.b` becomes `a\.b`, which never matches. This is
the classic heredoc/`%`-delimiter pattern: `<RegExpr String="&lt;&lt;(\w+)"
context="heredoc"/>` followed by `<StringDetect String="%1" dynamic="true"/>`.
Works when the captured token is `[A-Za-z0-9_]` only, breaks otherwise. 56
rules in 21 bundled files use dynamic `StringDetect`.

**How to implement.** Upstream's `replaceCaptures()`
(`upstream:src/lib/rule.cpp:89-96`) is the model:

```cpp
static QString replaceCaptures(const QString &pattern, const QStringList &captures, bool quote)
{
    auto result = pattern;
    for (int i = captures.size(); i >= 1; --i) {
        result.replace(QLatin1Char('%') + QString::number(i),
                       quote ? QRegularExpression::escape(captures.at(i - 1)) : captures.at(i - 1));
    }
    return result;
}
```

It is called with `quote = true` from `RegExpr`
(`upstream:src/lib/rule.cpp:696`) and `quote = false` from
`DynamicStringDetect` (`upstream:src/lib/rule.cpp:724`). Two details of that
loop matter and are easy to get wrong:

- It counts **down** from `captures.size()`, so `%10` is substituted before
  `%1` and a two-digit index is not mangled into `<capture 1> + "0"`.
- It starts at **1**, not 0. Upstream never fills capture group 0 into the
  list at all (`upstream:src/lib/rule.cpp:613-635`: *"we only need to compute
  the captured texts if we have real capture groups, highlightings should only
  address %1..%.., DetectChar ignores %0, too"*).

qutepart stores `match.capturedTexts()` (`src/hl/rules.cpp:348`), which
*includes* group 0 at index 0. That happens to make `%1`…`%4` line up
correctly — `data[N]` is group `N` — so the substitution itself is right, just
capped (§5.9). Port the escape flag while keeping qutepart's indexing:

```cpp
QString makeDynamicSubsctitutions(QString pattern, const QStringList &data, bool escape) {
    for (auto i = data.size() - 1; i >= 1; --i) {
        auto value = escape ? QRegularExpression::escape(data.at(i)) : data.at(i);
        pattern.replace(QLatin1Char('%') + QString::number(i), value);
    }
    return pattern;
}
```

Pass `true` from `RegExpRule`, `false` from `StringDetectRule`. This fixes
§5.9 at the same time. qutepart's extra `%0` (whole match) has no upstream
equivalent; no bundled file uses it, so it can simply be dropped.

## 1.4 Dynamic `DetectChar` reads the wrong capture group

**What it does.** `<DetectChar char="1" dynamic="true"/>` matches the first
character of **capture group 1** of the match that entered this context.
Unlike `%N` substitution, this one is a positional index, so the two
representations of "the captures list" must agree.

**What qutepart does.** They do not agree.
`DetectCharRule::tryMatchImpl()` (`src/hl/rules.cpp:202-222`):

```cpp
int matchIndex = this->index - 1;
...
pattern = textToMatch.contextData->at(matchIndex)[0];
```

`contextData` is `match.capturedTexts()`, which has the **whole match** at
index 0. So `char="1"` computes `matchIndex = 0` and reads the whole match
rather than group 1.

Upstream does the same `- 1` (`upstream:src/lib/rule.cpp:243-248`):

```cpp
m_captureIndex((data.dynamic ? data.char1.digitValue() : 0) - 1)
```

but its `captures` list *excludes* group 0
(`upstream:src/lib/rule.cpp:628-632` pushes groups `1..lastCapturedIndex`), so
`captures.at(0)` really is group 1.

8 rules in 2 bundled files use dynamic `DetectChar`: `perl.xml` and
`raku.xml` — heredoc and quote-delimiter handling, exactly where it is most
visible.

**How to implement.** Since qutepart keeps group 0, drop the `- 1`:

```cpp
int matchIndex = this->index;
```

and bounds-check against `contextData->length()` as it already does. Do this
*together* with §1.3 so both capture consumers agree on one indexing
convention — that is precisely the inconsistency that produced this bug.

---

# 2. Indentation-based folding

## 2.1 `<folding indentationsensitive="1"/>`

**What it does.** For languages with no explicit block delimiters — Python,
YAML, Haskell, CoffeeScript, Nim, Sass, reStructuredText — folding regions are
derived from leading whitespace rather than from `beginRegion`/`endRegion`
markers. A line starts a foldable region if the next non-empty line is indented
further; the region runs until indentation returns to the starting level.

**What qutepart does.** Nothing. `<general><folding>` is never read;
`loadLanguageSytnax()` (`src/hl/loader.cpp:662-708`) handles only `keywords`,
`comments` and `indentation` in its `<general>` loop (line 684). Folding
levels come exclusively from the
`beginRegion`/`endRegion` stack in `Context::parseBlock()`
(`src/hl/context.cpp:302-313`). 34 bundled files declare
`indentationsensitive`, and in all of them the fold column is simply empty.

**Note on prior art.** KSyntaxHighlighting does **not** implement the folding
algorithm — it only exposes the inputs, and KTextEditor/KatePart computes the
ranges. The upstream surface is exactly two calls:

- `Definition::indentationBasedFoldingEnabled()`
  (`upstream:src/lib/definition.h:319-327`), and
- `Definition::foldingIgnoreList()` (`upstream:src/lib/definition.h:330-340`),
  a `QStringList` of regexes for lines to treat as empty,

plus the per-state `State::indentationBasedFoldingEnabled()`
(`upstream:src/lib/state.cpp:79-85`), which answers for the **top context of
the current line's state** — that is how §2.2 is honoured. So there is no
algorithm to copy from upstream; qutepart has to write the folding logic
itself, and should mirror that API shape.

**How to implement.**

1. Parse it. Add a `folding` branch to the `<general>` loop in
   `loadLanguageSytnax()` (`src/hl/loader.cpp:684`) and carry a
   `bool indentationBasedFolding` on `Language` alongside the existing
   `indenter` field.

2. Compute levels. `TextBlockUserData::folding.level`
   (`src/text_block_user_data.h:27-30`) is already the single source of truth
   for the fold column (`src/side_areas.cpp:601-667`) and for
   `Qutepart::toggleFold()` (`src/qutepart.cpp:710-740`), so the change is
   localised: when the language is indentation-sensitive, set `folding.level`
   from the block's leading-whitespace width (in columns, tabs expanded)
   instead of from `data->regions.size()`.

3. Handle blank lines. A blank line must not close a region — it inherits the
   level of the next non-blank line. This is exactly what `<emptyLines>`
   (§5.2) is for, and why the two features are usually implemented together.

4. Re-highlighting. Folding level for line *N* depends on line *N+1*, so a
   change on one line can invalidate the fold level of the line above.
   qutepart's highlighter is a `QSyntaxHighlighter` walking forward; the
   simplest correct approach is to compute the level lazily in `FoldingArea`
   from the block's own indentation plus a forward peek, rather than caching a
   level the highlighter cannot keep consistent.

## 2.2 `noIndentationBasedFolding` on `<context>`

**What it does.** Suppresses §2.1 inside specific contexts — typically inside
multi-line strings and comments, where the text's indentation is content, not
structure. Meaningless until §2.1 exists; 10 bundled files set it.

**How to implement.** Parse it in `loadContext()` and store on `Context`.
Upstream combines the two flags at resolve time
(`upstream:src/lib/context_p.h:80-83`): the context answers "yes" only when the
definition enables indentation folding **and** the context does not set
`noIndentationBasedFolding`. Expose the same combined query on
`Qutepart::Context`, then have the folding-level computation ask the context
stack stored in `TextBlockUserData` for the line: if the top context says no,
reuse the previous line's level verbatim.

---

# 3. Missing rule and keyword attributes

## 3.1 `<include>` inside a keyword `<list>`

**What it does.** Composes keyword lists:

```xml
<list name="controlflow">
  <item>if</item>
  <include>keywords##Common</include>
</list>
```

`<include>listname</include>` splices in another list from the same file;
`<include>listname##Language</include>` splices one in from another language
definition. It is how TypeScript reuses JavaScript's lists rather than copying
them.

**What qutepart does.** Explicitly skips it, with a `qDebug()`
(`src/hl/loader.cpp:464-471`):

```cpp
if (xmlReader.name() == QLatin1String("include")) {
    // TODO includes in list is not supported yet
```

9 bundled files lose keywords this way, some badly:
`typescript.xml` (25 includes), `sass.xml` (17), `fortran-free.xml` (10),
`kateconfig.xml` (8), `raku.xml` (3), `selinux.xml` (3), plus `cabal.xml`,
`elvish.xml`, `fish.xml`. In `typescript.xml` this means most of the inherited
JavaScript keyword set is simply absent.

**How to implement.** Upstream's `KeywordList::resolveIncludeKeywords()`
(`upstream:src/lib/keywordlist.cpp:114-146`) is a compact model worth copying
rather than reinventing:

```cpp
while (!m_includes.isEmpty()) {
    const auto kw_include = std::move(m_includes.back());
    m_includes.pop_back();
    ...
    if (keywords) {
        if (this != keywords) {
            keywords->resolveIncludeKeywords(def);   // resolve target first
        }
        m_keywords += keywords->m_keywords;
    }
}
```

Three things to note:

- The pending-include list is **drained** (`pop_back`) as it is walked, so a
  cycle terminates naturally. No separate visited-set is needed; the only
  explicit guard is `this != keywords` for direct self-inclusion.
- Includes are resolved **recursively before** splicing, so a chain
  `A → B → C` gets all of C's words into A.
- For `list##Lang`, upstream calls `defData->load(OnlyKeywords(true))`
  (`upstream:src/lib/keywordlist.cpp:129-130`) — a partial load that parses the
  other definition's `<list>` elements without building its contexts. qutepart's
  `loadLanguage()` has no such mode; adding one is optional, but without it a
  single `##` include drags in a full language parse.

Concretely for qutepart:

1. In `loadKeywordList()` (`src/hl/loader.cpp:464-471`), collect `<include>`
   bodies into a parallel list instead of discarding them.
2. Resolve after `loadKeywordLists()` returns and **before**
   `context->setKeywordParams()` runs — `KeywordRule` snapshots the words into
   its own `items` hash at that point (`src/hl/rules.cpp:153-168`).
3. Cross-language includes need `Language` to expose its keyword lists, so keep
   the `QHash<QString, QStringList>` on `Language` rather than dropping it after
   `setKeywordParams()`.

Note the cross-language case interacts with case folding: the *including*
language's `casesensitive` decides how the borrowed words are stored, so splice
raw and let `makeKeywordsLowerCase()` run afterwards.

## 3.2 `IncludeRules includeAttrib`

**What it does.** `<IncludeRules context="X" includeAttrib="true"/>` makes the
including context also adopt `X`'s default `attribute` — i.e. text that no rule
matched is coloured with the *included* context's style, not the includer's.
Without it, unmatched text inside e.g. an embedded-language block falls back to
the host's colour, which typically shows up as stretches of "Normal Text" in
the middle of an embedded region.

**What qutepart does.** `loadIncludeRulesRule()` (`src/hl/loader.cpp:301-309`)
reads only `context`. 23 bundled files set `includeAttrib`.

**How to implement.** Upstream resolves this at load time, not match time
(`upstream:src/lib/context.cpp:59-130`): `Context::resolveIncludes()` walks
each `IncludeRules`, recursively resolves the target first, then **splices the
target's rules into this context's rule list**, erasing the `IncludeRules`
placeholder. The attribute is handled in the same loop:

```cpp
if (includeRules->includeAttribute()) {
    m_attributeFormat = context->m_attributeFormat;
}
```

Two semantics to match, both easy to get backwards:

- It is a plain assignment inside the loop, so with several `includeAttrib`
  includes in one context the **last one wins**, not the first.
- Resolution is a transitive closure — the target is resolved before being
  read, so an attribute included two levels down propagates up.

qutepart resolves `IncludeRules` by delegation at match time
(`IncludeRulesRule::tryMatchImpl()`) rather than by splicing, so it does not
need upstream's rule-list surgery. It only needs the attribute half: store the
flag on `IncludeRulesRule`, and in `Context::finalize()` — after
`resolveContextReferences()` has run, so the target is known — overwrite this
context's `style` with the included context's `style`, taking the last such
include in rule order.

## 3.3 `keyword insensitive` (per-rule)

**What it does.** Overrides the language-wide `<keywords casesensitive>` for a
single `<keyword>` rule. Used where most of a language is case-sensitive but
one class of words is not (built-in function names, preprocessor directives).

**What qutepart does.** `loadKeywordRule()` (`src/hl/loader.cpp:178-186`) reads
only `String`. `KeywordRule::setKeywordParams()` (`src/hl/rules.cpp:153-168`)
takes case sensitivity from the language and nothing else. 12 rules in 8
bundled files.

**How to implement.** Upstream never mutates the word list. `KeywordList`
keeps the original strings and lazily builds a *second* sorted lookup vector
for whichever sensitivity is asked for
(`upstream:src/lib/keywordlist.cpp:93-112`), and the rule picks between them:

```cpp
m_caseSensitivity(data.hasCaseSensitivityOverride ? data.caseSensitivityOverride
                                                  : keywordList.caseSensitivity())
```
(`upstream:src/lib/rule.cpp:520`)

qutepart takes the opposite approach — `makeKeywordsLowerCase()`
(`src/hl/loader.cpp:653-661`) rewrites the shared lists in place when the
language is case-insensitive. That makes a per-rule override awkward: a rule
wanting insensitive matching inside a case-*sensitive* language needs its own
folded copy. The smaller change is to stop the bulk mutation and let each
`KeywordRule` fold as it fills `items`:

```cpp
bool cs = insensitiveOverride.has_value() ? !(*insensitiveOverride) : newCaseSensitive;
this->caseSensitive = cs;
for (const auto &word : list) {
    items.insert(cs ? word : word.toLower(), true);
}
```

`tryMatchImpl()` already branches on `caseSensitive`, and `computeFirstChars()`
already walks `items`, so both keep working — but `computeFirstChars()` must
seed both cases of the first letter when the rule is insensitive, same caveat
as §1.2.

---

# 4. Styles and themes

## 4.1 `itemData backgroundColor` / `selBackgroundColor` / `selColor`

**What they do.** Per-attribute background colour, and the foreground/background
used when the text is selected. Backgrounds are how alert/TODO/error markers,
region markers and diff highlighting get their block colouring.

**What qutepart does.** `loadStyles()` (`src/hl/loader.cpp:510-575`) reads
`color` and `selColor` only, and never reads `backgroundColor` or
`selBackgroundColor`. `selColor` is then discarded anyway —
`makeFormat()` (`src/hl/style.cpp:168-169`) takes it as
`const QString & /*selColor*/`. 61 bundled files set `selColor`, 6 set
`backgroundColor`, 1 sets `selBackgroundColor`.

**How to implement.** Backgrounds are easy and worth doing:

```cpp
QString bgColor = getAttribute(attrs, "backgroundColor");
...
if (!bgColor.isNull()) {
    format->setBackground(QColor(bgColor));
}
```

The two selection colours are harder, because `QSyntaxHighlighter` formats and
`QPlainTextEdit`'s selection rendering are separate layers — the selection
palette wins. Implementing them properly means tracking a second
`QTextCharFormat` per style and swapping it in `Qutepart`'s
`ExtraSelection` handling for selected ranges. Given that the visual payoff is
small, parsing-and-ignoring is a defensible choice — but it should be an
explicit comment rather than an unused parameter.

## 4.2 Theme `custom-styles` is parsed but never applied

**What it does.** A `.theme` file's `custom-styles` block overrides colours per
*language and attribute*, not just per default style:

```json
"custom-styles": {
  "ISO C++": { "Keyword": { "text-color": "#c678dd", "bold": true } }
}
```

This is how Breeze, Dracula, Solarized et al. fine-tune individual languages.

**What qutepart does.** `Theme::load()` (`src/theme.cpp:43-62`) parses the
block into `Theme::customStyles` (`include/qutepart/theme.h:94`). Nothing ever
reads it — there is no getter, and `applyTheme()` (`src/hl/style.cpp:19-67`)
consults only `theme->getTextStyles()`. Every bundled theme's per-language
tuning is inert.

**How to implement.** `Style` already knows its `defStyleName` and gets
`setTheme()` called on it; it needs two more pieces of context — the language
name and the itemData name — to look up `customStyles[lang][itemData]`.

1. Keep the itemData `name` on `Style` (currently only `defStyleName` is kept;
   `loadStyles()` has the name and drops it).
2. `Context::setTheme()` (`src/hl/context.cpp:53`) already walks styles with
   the language in scope — pass `language->name()` down.
3. In `applyTheme()`, apply `getTextStyles()[defStyle]` first, then overlay
   `customStyles[langName][itemDataName]` using the same key loop.

Add a `getCustomStyles()` accessor on `Theme` while doing this.

## 4.3 `dsOperator` has no default format

`defaultFormat()` (`src/hl/style.cpp:85-87`) has an empty `// TODO` branch for
`dsOperator`, so operators render as plain text in the built-in palette.
Upstream Breeze gives it a distinct colour (`#4f97d7` dark / `#1f1c1b` light).
Same for `dsConstant`, which is bold-only here. These only surface when no
`.theme` is loaded, since a theme's `text-styles` will override them — but the
no-theme path is the default one.

---

# 5. Smaller gaps

## 5.1 `<comment position>`

```xml
<comment name="singleLine" start="#" position="afterwhitespace"/>
```

`position="afterwhitespace"` tells the comment-toggle action to insert the
marker after the leading indentation rather than in column 0. Upstream exposes
it as `Definition::singleLineCommentPosition()` returning a
`CommentPosition` enum (`upstream:src/lib/definition.cpp:726-731`).

`loadComments()` (`src/hl/loader.cpp:580-611`) reads `name`, `start` and `end`
only. 30 bundled files set `position`. Cheap: store it on `Language` next to
the existing `commentStart`/`commentEnd`/`commentSingleLine` fields and consult
it in whatever drives line-commenting.

**Not a gap: `<comment region>`.** The XSD allows it and 57 bundled files set
it, but upstream's `loadComments()` reads only `name`, `start`, `end` and
`position` — `region` is never parsed there either. It is a schema-only
attribute; ignoring it matches upstream exactly.

Worth fixing while in this function: `getAttribute(attrs, "end", error)` at
`src/hl/loader.cpp:598` passes `error` as the *defaultValue* parameter.
Harmless while `error` is null, but it is not what was meant.

## 5.2 `<emptyLines>` / `<emptyLine regexpr="...">`

Defines which lines count as "empty" for folding purposes beyond literally
blank ones — e.g. a line containing only a comment marker, or only whitespace
and a `\`. Required for §2.1 to behave on real code. 9 bundled files declare
it; qutepart's `<general>` loop ignores the element entirely, and
`Context::parseBlock()` (`src/hl/context.cpp:290`) treats only a
zero-length line as empty for `lineEmptyContext`.

Upstream stores these as `DefinitionData::foldingIgnoreList`, a plain
`QStringList` of regex *sources* handed to the caller
(`upstream:src/lib/definition.cpp:752-775`), and reads only the `regexpr`
attribute — the `casesensitive` attribute the XSD allows on `<emptyLine>` is
**not** parsed upstream either, so it can be skipped.

Implement as a `QList<QRegularExpression>` on `Language`, consulted by both the
`lineEmptyContext` check and the indentation-folding level computation.

## 5.3 Spell-checking

Two halves, both absent:

- `<spellchecking><encodings><encoding string="&amp;auml;" char="ä"/></encodings></spellchecking>`
  maps entity-style escapes to real characters so the spell checker sees words,
  not markup. 2 bundled files (the TeX/LaTeX family) use it.
- `itemData spellChecking="false"` marks an attribute as not spell-checkable —
  code, identifiers, URLs. This one *is* parsed (`src/hl/loader.cpp:544`) into
  the `flags` hash, but `makeFormat()` (`src/hl/style.cpp:168-203`) only looks
  at `italic`/`bold`/`underline`/`strikeout` and drops it.

Neither matters until qutepart grows a spell checker. When it does, the cheap
half is `spellChecking`: keep it as a `bool` on `Style` next to `_textType` —
`Style::textType()` is already the mechanism by which the editor asks "is this
comment or string text?", so a `spellCheckable()` sibling fits the existing
shape exactly.

## 5.4 `<keywords wordWrapDeliminator>`

Characters at which soft word-wrap may break a long line. 9 bundled files.
Note the upstream semantics (`upstream:src/lib/definition.cpp:678-683`): the
wrap set **defaults to the word-delimiter set** and the attribute *appends* to
it — it does not replace it:

```cpp
if (wordWrapDeliminatorAttr.isEmpty()) {
    wordWrapDelimiters = wordDelimiters;
} else {
    wordWrapDelimiters.append(wordWrapDeliminatorAttr);
}
```

Parse in `loadKeywordParams()`, store on `Language`, consult from the wrapping
code. Low value unless soft wrap is implemented.

## 5.5 `LineContinue char`

`<LineContinue char="&amp;"/>` — the continuation character, defaulting to `\`.
Fortran (`&`), some assembler dialects and VB (`_`) need the override.
`loadRule()` (`src/hl/loader.cpp:351`) constructs `new LineContinueRule(params)`
with no arguments, and `LineContinueRule::tryMatchImpl()` hard-codes the
backslash. 18 rules in 7 bundled files. Trivial: add a `QChar` to the
constructor, default `'\\'`.

## 5.6 Per-rule `weakDeliminator` / `additionalDeliminator`

The XSD allows these on `keyword`, `Int`, `Float`, `HlCOct`, `HlCHex` and
`WordDetect`, not just on `<general><keywords>`. qutepart supports the
`<keywords>` form only (`loadKeywordParams()`). Only 2 bundled files
(`klipper-config.xml`, `twig.xml`, 3 rules total) use the per-rule form, so
this is genuinely low priority — but note that when it is implemented it must
go through `sharedDeliminatorSet()` so the per-rule set is still interned.

## 5.7 `<language>` metadata attributes

Not read at all: `alternativeNames` (5 bundled files), `section`, `author`,
`license`, `style`, `generated`. `parseXmlFile()` reads `name`, `extensions`,
`mimetypes`, `priority`, `hidden`, `indenter`.

`alternativeNames` is the one with behaviour attached — it is how
`##Language` external-context references and user-facing language selection
resolve aliases (`loadExternalContext()`, `src/hl/loader.cpp:865-899`, and
`chooseLanguage()` in `language_db.cpp`). `section` is the menu grouping for a
language picker. The rest are informational.

## 5.8 `stopEmptyLineContextSwitchLoop`, and the empty-line switch loop

**What it does.** On an empty line, upstream does not switch
`lineEmptyContext` once — it switches **repeatedly**, until the top context has
no `lineEmptyContext`, or sets `stopEmptyLineContextSwitchLoop`, or a `#pop`
hits the main context, or a 1024-iteration guard trips
(`upstream:src/lib/abstracthighlighter.cpp:144-166`):

```cpp
while (!stateData->topContext()->lineEmptyContext().isStay()) {
    if (!d->switchContext(stateData, stateData->topContext()->lineEmptyContext(), ...)) {
        break;   // tried to #pop the main context
    }
    if (stateData->topContext()->stopEmptyLineContextSwitchLoop()) {
        break;
    }
    ...
}
```

The attribute exists to stop that loop one step early where a definition wants
a single hop. It was added for [D18509](https://phabricator.kde.org/D18509).

**What qutepart does.** `Context::parseBlock()` (`src/hl/context.cpp:290-293`)
switches exactly once and returns:

```cpp
if (textToMatch.isEmpty() && (!_lineEmptyContext.isNull())) {
    contextStack.switchTo(_lineEmptyContext);
    return;
}
```

So the gap is not the attribute — it is the missing loop. Chained
`lineEmptyContext` definitions unwind one level per empty line instead of
settling immediately, which shows up as blank lines inside nested constructs
holding a stale context. No bundled file sets
`stopEmptyLineContextSwitchLoop`, so implementing the loop matters more than
implementing the attribute; add both together, since the loop without the
brake is the thing the attribute exists to bound.

**Unrelated, same area.** `loadContext()` reads `lineBeginContext` from the
`"lineEndContext"` attribute (`src/hl/loader.cpp:403`, a copy-paste of line
401). `lineBeginContext` is not in the Kate schema and upstream has no such
concept; since the `while` loop in `Language::switchAtEndOfLine()` guarantees
the extra switch at `src/hl/language.cpp:143-145` is a no-op, this is dead code
rather than a live bug — but it should be deleted rather than left as a trap.

## 5.9 Dynamic captures capped at `%4`

`makeDynamicSubsctitutions()` (`src/hl/rules.cpp:124-131`) handles `%0` through
`%4`. Upstream substitutes `%1` through `%N` for however many capture groups
the entering match produced — there is no fixed ceiling
(`upstream:src/lib/rule.cpp:89-96`). 5 bundled rules reference `%5` or higher
and silently keep the literal `%5` in their pattern.

`dropDynamicWithoutPlaceholders()` (`src/hl/rules.cpp:111-121`) already scans
for `%` + any digit, so it correctly keeps those rules marked dynamic — only
the substitution table is short. Fixed by the descending loop in §1.3, which
also handles the two-digit `%10`+ case that a naive ascending loop corrupts.

## 5.10 MIME types are read from the wrong attribute, then overwritten

Two independent bugs stacked on each other.

First, the attribute name. Upstream reads **`mimetype`**, singular
(`upstream:src/lib/definition.cpp:466`):

```cpp
const auto mts = obj.value(QLatin1String("mimetype")).toString();
```

qutepart reads `mimetypes`, plural (`src/hl/loader.cpp:755`). Across the 385
bundled files, `mimetype=` appears 275 times and `mimetypes=` **zero** times —
so `mimetypesStr` is null for every single language.

Second, even that null is discarded. `src/hl/loader.cpp:776`:

```cpp
QStringList extensions = extensionsStr.split(';', Qt::SplitBehaviorFlags::SkipEmptyParts);
QStringList mimetypes  = extensionsStr.split(';', Qt::SplitBehaviorFlags::SkipEmptyParts);
```

Every `Language` therefore reports its glob patterns (`*.cpp;*.cc;…`) as its
MIME types. `chooseLanguage()` takes a `mimeType` parameter, so anything
selecting a language by MIME type is matching globs against MIME strings and
never hitting. Read `mimetype` (accepting `mimetypes` as an alias) and split
*that*.

---

# 6. Syntax data freshness

Independent of the parser, the bundled definitions are old. Highest
`kateversion` in `syntax/` is 5.79; upstream is on the 6.x series.

Version drift on common languages (`version=` attribute, local vs upstream
master as of 2026-08-25):

| Language | bundled | upstream |
|----------|---------|----------|
| markdown.xml | 3 | 33 |
| python.xml | 6 | 33 |
| javascript.xml | 8 | 26 |
| cpp.xml | 6 | 21 |
| html.xml | 5 | 20 |
| yaml.xml | 3 | 17 |
| rust.xml | 5 | 16 |
| java.xml | 4 | 13 |
| go.xml | 4 | 10 |
| typescript.xml | 18 | 18 |

37 upstream languages are absent entirely:

`abnf`, `acpi-asl`, `acpi-dsl`, `asm-arm`, `bnf`, `capnproto`, `d2`, `dts`,
`edp`, `elixir-eex`, `elixir-heex`, `expect`, `gemtext`, `gn`,
`gtk-blueprint`, `hjson`, `jinja`, `json5`, `just`, `latex-logfile`,
`logfile-advanced-selector`, `logfile-selector`, `mermaid`, `nftables`,
`pioasm`, `quarto`, `sas`, `slint`, `snakemake`, `sparql`, `ssim`,
`starlark`, `step`, `toit`, `webvtt`, `xhtml`, `xkb`.

12 files are local-only and would need to survive any bulk re-sync:
`alert_indent`, `asterisk`, `chicken`, `css-php`, `doxygenlua`, `fortran`,
`html-php`, `javascript-php`, `varnish4`, `varnishcc4`, `varnishtest4`,
`xslt`.

**How to approach it.** Re-syncing before fixing §1 and §3 is the wrong order —
newer definitions use the newer attributes more heavily, so pulling them in
first would make highlighting worse, not better. Suggested sequence: fix §1
(correctness), then §3.1/§3.2 (includes), then re-sync, then §2 (folding).

`regenerate-language-db.py` and `regenerate-qrc.py` already exist to rebuild
`language_db_generated.cpp` and the `.qrc` after a sync. A re-sync should be
scripted against a pinned upstream tag so it is repeatable, with the 12
local-only files listed as explicit keeps.

---

# 7. Indenters

`<language indenter="...">` names a Kate indenter. Upstream ships them as
JavaScript; qutepart reimplements them in C++
(`INDENT_ALG_*` in `src/indent/indenter.cpp:52-86`).

Implemented (8): `none`, `normal`, `lisp`, `xml`, `scheme`, `python`, `ruby`,
`markdown`, `cstyle`.

Referenced by bundled syntax files but **not** implemented (7 modes, 12 files):
`haskell` (5 files), `ada` (2), `pli`, `pascal`, `lua`, `lilypond`, `elixir`.
Those files fall back to `normal`.

Also shipped upstream and not present here (not referenced by any bundled
file, so lower priority): `cmake`, `cppstyle`, `django`, `latex`, `julia`,
`R`, `replicode`.

`haskell` is the one worth doing — 5 languages, and it pairs naturally with
§2.1 since Haskell is also indentation-folded.

---

# Appendix: what is *not* missing

Worth stating explicitly, since it is the bulk of the format:

- **All 18 rule types** are implemented: `keyword`, `Float`, `HlCOct`,
  `HlCHex`, `Int`, `DetectChar`, `Detect2Chars`, `AnyChar`, `StringDetect`,
  `WordDetect`, `RegExpr`, `LineContinue`, `HlCStringChar`, `RangeDetect`,
  `HlCChar`, `IncludeRules`, `DetectSpaces`, `DetectIdentifier`.
- **All 31 `dsXxx` default styles** are recognised by `defaultFormat()`
  (`src/hl/style.cpp:68-141`), matching the XSD's enumeration exactly.
- **Common rule attributes** all work: `attribute`, `context`, `lookAhead`,
  `firstNonSpace`, `column`, `dynamic`, `beginRegion`, `endRegion`.
- **Context switching** is complete: `#stay`, `#pop`, repeated `#pop`,
  `#pop!Context`, and cross-language `Context##Language` / `##Language`.
- **Region-marker folding** works, as does `lineEndContext` and dynamic
  contexts.
- **Theme `text-styles` and `editor-colors`** are parsed and applied.
- **`%1`…`%4` capture substitution is correctly indexed**, despite qutepart and
  upstream using different capture-list representations (§1.3). Only
  `DetectChar` gets the indexing wrong (§1.4).

qutepart-cpp also has machinery upstream does not: the per-character rule
bucketing in `Context::buildRuleIndex()` and the `FirstCharSet` precomputation,
which is what makes it fast. Every fix above needs to keep
`computeFirstChars()` honest — a rule whose match set widens (case-insensitive
matching, spliced keyword lists) must widen its first-char set to match, or the
new matches will never be attempted.
